### ISSUE TITLE
Tweedie loss 

### ISSUE THREAD

**Issue by kaleming** [awslabs/gluon-ts#984]
Does gluon-ts provide provide DeepAR with Tweedie loss ?

## Tweedie distribution

![tweedie-distribution](https://user-images.githubusercontent.com/51836477/89934535-1b4a3d80-dbe7-11ea-8082-685fd52c2b0f.png)

-   It is a type exponental dispersion model (ED). ED is a very general type of probability distribution models where the PDF (probab. distribution function) is represented using exponents. A very common example would be the normal distrbution whose PDF is well known (and exponential).
-   It has a special relation between mean and variance. i.e. the variance ~ some power of mean (Variance=k*Mean^p). This constant, "p" is important and is the tweedie_variance_power parameter in the lightgbm's API (default value 1.5).
-  The well known poisson (p=1) and gamma (p=2) distributions can be considered special cases of tweedie. (Now we know why the kernels using poisson regression works better than L2 regression and tweedie works even better)
-   The distribution is not defined for p>2 and p<1

The data in tweedie distribution is "concentrated" (technically speaking, has a high point mass) at 0. Here's a plot of tweedie distribution:

The loss function of tweedie is created for the aim of maximizing the negative log likelihood of the tweedie distribution. Lightgbm computes tweedie loss as follows:

### Tweedie loss

The loss function of tweedie is created for the aim of maximizing the negative log likelihood of the tweedie distribution. Lightgbm computes tweedie loss as follows: 

```
def tweedie_eval(y_pred, y_true, p=1.5):
    y_true = y_true.get_label()
    a = y_true*np.exp(y_pred, (1-p)) / (1-p)
    b = np.exp(y_pred, (2-p))/(2-p)
    loss = -a + b
    return loss 
```

Source
[Kaggle](https://www.kaggle.com/c/m5-forecasting-accuracy/discussion/150614)
## References
- list known implementations:
[3th place kaggle](https://www.kaggle.com/c/m5-forecasting-accuracy/discussion/164374)

**Comment by lostella** [comment]
@kaleming Tweedie distributions are not implemented in GluonTS, but I agree that this would be really cool to have. Is this something you would be interested in contributing with a PR?

**Comment by kaleming** [comment]
Hi @lostella, 
I will have a look on MxNet docs and see if I can help on it.

**Comment by statgirl30** [comment]
just wanted to mention that tweedie loss is -2*logloss. so you should probably multiply it by 2?

**Comment by timoschowski** [comment]
wanted to check-in on this, since there was some interested voiced on this at ISF. @kaleming is this something that you were able to make progress on?

**Comment by kaleming** [comment]
Hi @timoschowski , 

Sorry my late reply @statgirl30. I was busy these last months, and I couldn't handle to learn mxnet and pydantic/pytest to convert this code of [tweedie](https://github.com/thequackdaddy/tweedie/blob/master/tweedie/tweedie_dist.py) into gluon-ts format.

Please feel free guys to go ahead on this feature request.

**Comment by timoschowski** [comment]
This is mainly for self-documentation as I didn't make much progress looking into this. 

The Tweedie loss comes from: 
* https://arxiv.org/pdf/1508.06378.pdf equation (10) and (11) are the key equations. This assumes the power parameter to be given (set to 1.5 typically, it seems) and same for the rate parameter phi 
* Section 4.2 of the above paper discusses how to estimate (p, phi) via profile likelihood
* LightGBM uses log of the input in its implementation https://github.com/microsoft/LightGBM/blob/1c27a15e42f0076492fcc966b9dbcf9da6042823/src/metric/regression_metric.hpp#L300-L318
* a comprehensive python package that also contains sampling and other methods is here: https://github.com/thequackdaddy/tweedie

**Comment by timoschowski** [comment]
https://github.com/awslabs/gluonts/pull/3277 this is a PR using Claude.
